[CI] CardinalityWithRequestBreakerIT.testRequestBreaker fails on 7.9

**Build scan**:

https://gradle-enterprise.elastic.co/s/2coyhtowa22hi

**Repro line**:

./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.search.aggregations.metrics.CardinalityWithRequestBreakerIT.testRequestBreaker" \
  -Dtests.seed=ABEA4F928B9BAFCF \
  -Dtests.security.manager=true \
  -Dtests.locale=hr \
  -Dtests.timezone=Asia/Ust-Nera \
  -Druntime.java=8

**Reproduces locally?**:

yes

**Applicable branches**:

7.9

**Failure history**:

Another one on Sep 8th might be related: https://gradle-enterprise.elastic.co/s/e2ht4jwbic3dq

**Failure excerpt**:

```
java.lang.AssertionError: Request breaker not reset to 0 on node: node_s1
    Expected: <0L>
         but: was <524288L>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.test.InternalTestCluster.lambda$ensureEstimatedStats$39(InternalTestCluster.java:2501)
        at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:951)
        at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:924)
        at org.elasticsearch.test.InternalTestCluster.ensureEstimatedStats(InternalTestCluster.java:2499)
        at org.elasticsearch.test.TestCluster.assertAfterTest(TestCluster.java:94)
        at org.elasticsearch.test.InternalTestCluster.assertAfterTest(InternalTestCluster.java:2523)
        at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:598)
        at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2233)
```
```
java.lang.RuntimeException: 2 arrays have not been released
        at org.elasticsearch.common.util.MockBigArrays.ensureAllArraysAreReleased(MockBigArrays.java:70)
        at org.elasticsearch.test.ESTestCase.checkStaticState(ESTestCase.java:528)
        at org.elasticsearch.test.ESTestCase.after(ESTestCase.java:368)
```

--- a/server/src/main/java/org/elasticsearch/search/aggregations/metrics/HyperLogLogPlusPlus.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/metrics/HyperLogLogPlusPlus.java
@@ -78,9 +78,23 @@ public final class HyperLogLogPlusPlus extends AbstractHyperLogLogPlusPlus {
 
     public HyperLogLogPlusPlus(int precision, BigArrays bigArrays, long initialBucketCount) {
         super(precision);
-        hll = new HyperLogLog(bigArrays, initialBucketCount, precision);
-        lc = new LinearCounting(bigArrays, initialBucketCount, precision, hll);
-        algorithm = new BitArray(1, bigArrays);
+        HyperLogLog hll = null;
+        LinearCounting lc = null;
+        BitArray algorithm = null;
+        boolean success = false;
+        try {
+            hll = new HyperLogLog(bigArrays, initialBucketCount, precision);
+            lc = new LinearCounting(bigArrays, initialBucketCount, precision, hll);
+            algorithm = new BitArray(1, bigArrays);
+            success = true;
+        } finally {
+            if (success == false) {
+                Releasables.close(hll, lc, algorithm);
+            }
+        }
+        this.hll = hll;
+        this.lc = lc;
+        this.algorithm = algorithm;
     }
 
     @Override

--- a/server/src/test/java/org/elasticsearch/search/aggregations/metrics/HyperLogLogPlusPlusTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/metrics/HyperLogLogPlusPlusTests.java
@@ -21,12 +21,21 @@ package org.elasticsearch.search.aggregations.metrics;
 
 import com.carrotsearch.hppc.BitMixer;
 import com.carrotsearch.hppc.IntHashSet;
+import org.elasticsearch.common.breaker.CircuitBreaker;
+import org.elasticsearch.common.breaker.CircuitBreakingException;
+import org.elasticsearch.common.breaker.NoopCircuitBreaker;
 import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.indices.breaker.CircuitBreakerService;
 import org.elasticsearch.test.ESTestCase;
+
+import java.util.concurrent.atomic.AtomicLong;
 
 import static org.elasticsearch.search.aggregations.metrics.AbstractHyperLogLog.MAX_PRECISION;
 import static org.elasticsearch.search.aggregations.metrics.AbstractHyperLogLog.MIN_PRECISION;
 import static org.hamcrest.Matchers.closeTo;
+import static org.hamcrest.Matchers.equalTo;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 public class HyperLogLogPlusPlusTests extends ESTestCase {
     public void testEncodeDecode() {
@@ -125,5 +134,41 @@ public class HyperLogLogPlusPlusTests extends ESTestCase {
         assertEquals(16, HyperLogLogPlusPlus.precisionFromThreshold(10000));
         assertEquals(18, HyperLogLogPlusPlus.precisionFromThreshold(100000));
         assertEquals(18, HyperLogLogPlusPlus.precisionFromThreshold(1000000));
+    }
+
+    public void testCircuitBreakerOnConstruction() {
+        int whenToBreak = randomInt(10);
+        AtomicLong total = new AtomicLong();
+        CircuitBreakerService breakerService = mock(CircuitBreakerService.class);
+        when(breakerService.getBreaker(CircuitBreaker.REQUEST)).thenReturn(new NoopCircuitBreaker(CircuitBreaker.REQUEST) {
+            private int countDown = whenToBreak;
+            @Override
+            public double addEstimateBytesAndMaybeBreak(long bytes, String label) throws CircuitBreakingException {
+                if (countDown-- == 0) {
+                    throw new CircuitBreakingException("test error", bytes, Long.MAX_VALUE, Durability.TRANSIENT);
+                }
+                total.addAndGet(bytes);
+                return total.get();
+            }
+
+            @Override
+            public long addWithoutBreaking(long bytes) {
+                total.addAndGet(bytes);
+                return total.get();
+            }
+        });
+        BigArrays bigArrays = new BigArrays(null, breakerService, CircuitBreaker.REQUEST).withCircuitBreaking();
+        final int p = randomIntBetween(HyperLogLogPlusPlus.MIN_PRECISION, HyperLogLogPlusPlus.MAX_PRECISION);
+        try {
+            for (int i = 0; i < whenToBreak + 1; ++i) {
+                final HyperLogLogPlusPlus subject = new HyperLogLogPlusPlus(p, bigArrays, 0);
+                subject.close();
+            }
+            fail("Must fail");
+        } catch (CircuitBreakingException e) {
+            // OK
+        }
+
+        assertThat(total.get(), equalTo(0L));
     }
 }

